src/protocol/boolean/random_bits_generator.rs

```diff
@@ -0,0 +1,140 @@
+/// which makes the `buffer` hash map a ring buffer.
+#[allow(dead_code)]
+pub struct RandomBitsGenerator<'a, F: Field> {
+    context: ProtocolContext<'a, Replicated<F>, F>,
+    buffer: HashMap<i8, RandomBitsShare<F>>,
+    counter: u32,
+    read_pointer: i8,
+    write_pointer: i8,
+    abort_count: u32,
+}
+
+#[allow(dead_code)]
+impl<'a, F: Field> RandomBitsGenerator<'a, F> {
+    const REFILL_THRESHOLD: u8 = 16;
+
+    pub fn new(context: ProtocolContext<'a, Replicated<F>, F>) -> Self {
+        Self {
+            context,
+            buffer: HashMap::with_capacity(u8::MAX.into()),
+            counter: 0,
+            read_pointer: 0,
+            write_pointer: 0,
+            abort_count: 0,
+        }
+    }
+
+    /// Takes one `RandomBitsShare` instance out of the buffer. If the number
+    /// of buffered items fall below a threshold, it'll replenish.
+    pub async fn take_one(&mut self) -> Result<RandomBitsShare<F>, Error> {
+        self.refill_if_needed().await?;
+        let r = self
+            .buffer
+            .remove(&self.read_pointer)
+            .expect("unexpected buffer depletion");
+        self.read_pointer = self.read_pointer.wrapping_add(1);
+        Ok(r)
+    }
+
+    /// Checks if the stock is short, and replenish as needed. Specifically, it
+    /// will call `SolvedBits` once the stock falls below `REFILL_THRESHOLD`,
+    /// until the `write_pointer` is one place behind the `read_pointer`.
+    async fn refill_if_needed(&mut self) -> Result<(), Error> {
+        // intentional type coercing `i8` -> `u8`
+        #[allow(clippy::cast_sign_loss)]
+        if self.write_pointer.wrapping_sub(self.read_pointer) as u8 > Self::REFILL_THRESHOLD {
+            return Ok(());
+        }
+
+        while self.read_pointer.wrapping_sub(self.write_pointer) != 1 {
+            let record_id = RecordId::from(self.counter);
+            let r = SolvedBits::execute(self.context.bind(record_id), record_id).await?;
+            if let Some(x) = r {
+                self.buffer.insert(self.write_pointer, x);
+                self.write_pointer = self.write_pointer.wrapping_add(1);
+            } else {
+                // Keep track of how many times `SolvedBits` aborted. This stat
+                // is for testing and telemetry purposes only.
+                self.abort_count += 1;
+            }
+            // Current implementation will cause the context to panic once u32
+            // wraps around. It's probably safe to reuse the same index again
+            // after so many records being processed.
+            self.counter = self.counter.wrapping_add(1);
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::RandomBitsGenerator;
+    use crate::{
+        error::Error,
+        ff::Fp31,
+        protocol::QueryId,
+        test_fixture::{join3, logging, make_contexts, make_world, TestWorld},
+    };
+
+    #[tokio::test]
+    pub async fn basic() -> Result<(), Error> {
```

Comment:
This test takes about 6 seconds on my machine to complete because it generates about 500 bit shares. Would be great if I can mock `SolvedBits` to return bogus outputs. Mocking is something we'll need to support soon.